### ISSUE TITLE
Flowable does not use the configured Datasource

### ISSUE THREAD

**Issue by anujjamwal** [flowable/flowable-engine#1902]
I am trying to setup flowable in a spring boot with two data sources. One for flowable and other for the application JPA. I can see the correct datasource being assigned in the SpringProcessEngineConfiguration but app startup fails with error:

java.lang.IllegalArgumentException: dataSource or dataSourceClassName or jdbcUrl is required.
	at com.zaxxer.hikari.HikariConfig.validate(HikariConfig.java:955) ~[HikariCP-3.2.0.jar:na]

**properties file**

`
flowable.datasource.jdbc-url=jdbc:h2:mem:testdb
flowable.datasource.driverClassName=org.h2.Driver
flowable.datasource.username=sa
flowable.datasource.password=password
flowable.idm.enabled=false
spring.h2.console.enabled=true

spring.data.mongodb.host=localhost
spring.data.mongodb.port=27017
spring.data.mongodb.database=database
`

**Below is the datasource configuration**
`
@Configuration
public class DataSourceConfigurer {
  @Bean
  @Primary
  @ConfigurationProperties("flowable.datasource")
  public DataSourceProperties dataSourceProperties() {
    return new DataSourceProperties();
  }

  @Bean(name = "flowableDataSource")
  @ConfigurationProperties(prefix = "flowable.datasource")
  public DataSource flowableDataSource(DataSourceProperties properties) {
    return properties.initializeDataSourceBuilder().build();
  }

  @Bean
  @Primary
  @ConfigurationProperties(prefix = "spring.datasource")
  public DataSource primaryDataSource() {
    return DataSourceBuilder
      .create()
      .build();
  }

}`

**I added the following configuration.**

`@Configuration
@AutoConfigureAfter(ProcessEngineAutoConfiguration.class)
public class FlowableConfiguration {
  @Autowired
  @Qualifier("flowableDataSource")
  private DataSource dataSource;

  @Bean
  public EngineConfigurationConfigurer<SpringProcessEngineConfiguration> processEngineConfigurer() {
    return configuration -> {
      configuration.setDataSource(dataSource);
    };
  }
}`

**Comment by wuwu2000** [comment]
Hello, is there any update on this?
I have a similar problem. My configuration is

```
@Bean
  public SpringProcessEngineConfiguration processEngineConfiguration(ResourceLoader resourceLoader,
                                                                     @Qualifier("bpmnDataSource") DataSource dataSource,
                                                                     @Qualifier("bpmnTransactionManager") PlatformTransactionManager transactionManager,
                                                                     ConfigurationService configurationService  ) {

    final SpringProcessEngineConfiguration engineConfiguration = new SpringProcessEngineConfiguration();

    engineConfiguration.setFlowable5CompatibilityEnabled(true);
    engineConfiguration.setFlowable5CompatibilityHandler(new DefaultFlowable5SpringCompatibilityHandler());

    engineConfiguration.setDataSource(dataSource);
    engineConfiguration.setTransactionManager(transactionManager);
    engineConfiguration.setDatabaseSchemaUpdate("true");

    engineConfiguration.setDeploymentResources(new Resource[] { resourceLoader.getResource("classpath:workflow2.bpmn") });
    engineConfiguration.setHistory(HistoryLevel.FULL.getKey());

    Map<Object, Object> beans = new HashMap<>();
    beans.put("config", configurationService);
    engineConfiguration.setBeans(beans);

    return engineConfiguration;
  }
```

I debugged `org.flowable.common.engine.impl.interceptor.CommandContext#getSession`
Most of the time this method returns a session bound to the given dataSource but at some point this returns a session bound to a different dataSource, so it cannot find the tables of course. Those are the last log entries until exception

```
INFO  [org.flowable.cmmn.engine.impl.CmmnEngineImpl]:72   CmmnEngine default created
ERROR [org.flowable.common.engine.impl.db.CommonDbSchemaManager]:204   Could not get property from table ACT_GE_PROPERTY
```

**Comment by wuwu2000** [comment]
I don't know why it's needed but a workaround is to create an additional configuration class which configures `AppEngineFactoryBean `


```
@Configuration
public class FlowableConfigurator implements InitializingBean {

    @Autowired
    @Qualifier("bpmnDataSource")
    private DataSource dataSource;

    @Autowired
    private AppEngineFactoryBean appEngineFactoryBean;


    @Override
    public void afterPropertiesSet()
            throws Exception {

        appEngineFactoryBean.getAppEngineConfiguration().setDataSource(dataSource);
    }
}
```

**Comment by wuwu2000** [comment]
@filiphr so I can drop all my configurations to SpringProcessEngineConfiguration but configure AppEngineFactoryBean?

**Comment by filiphr** [comment]
Yes indeed, you should only need to configure the `AppEngineFactoryBean`
